### ISSUE TITLE
Bulk application deployment

### ISSUE THREAD

**Issue by llopat** [nmaas-platform/nmaas-platform#17]
The domain administrator should have the possibility to trigger a deployment of several application instances within a particular domain with a single call.
Specification of the list of application instances to be deployed should be provided in a csv file (exact schema to be designed).

The platform should expose additional API endpoint to trigger the deployment.
A comprehensive validation of the request needs to be performed (parsing of the csv content and validation against the content of the internal database like application names, subscriptions and so on) and meaningful information should be returned if any validation errors occur.

**Comment by llopat** [comment]
@korvoj could be please propose the csv format

**Comment by korvoj** [comment]
I propose to go through several iterations/discussions so that we can make the schema as versatile as possible, applicable to different scenarios. 

## Core Principles

I propose the following core principles when dealing with the schema:
- define required and optional fields. Optional fields if not specified, take some default value (to be determined on a case-by-case basis)
- accept CSV files which include fields that have not been defined, and ignore them on purpose. Perhaps a warning message can be shown to the user, but this is optional. The idea is to allow users to deploy applications in bulk by creating CSV dumps from different systems. It will be easier if only the necessary columns are added/renamed instead of tailoring the CSV export to the same exact format that NMaaS requires.
- we should evaluate how hard it would be (from a technical perspective) to implement dynamic mapping of fields. This is something that Moodle does, a screenshot is available below, to get a better idea. The main concept is that CSV upload is done in two steps. The first step is the upload of the file itself, while the second steps allows the user to specify what fields in the CSV file correspond to what expected values. A dropdown field is available for each field detected in the CSV file, allowing it to be mapped, if needed (there is also a possibility to simply ignore its value). This is just an idea, we might conclude that the extra effort is not worth it...

![ss](https://user-images.githubusercontent.com/4062681/219037808-571a34c5-3228-4f91-b6bf-8eee14f0aadd.png)

## Initial CSV Schema

Regarding the initial version of the proposal, I suggest the following fields to be parseable from the CSV file:
- `domain name` [Required] - the name of the domain where the application needs to be deployed
- `application name` [Required] - the name of the application (as visible from the catalog) to be deployed
- `application version` [Optional] - the version of the application to be deployed. If not specified, the latest version available in the catalog is deployed.
- `...parameters` [Optional] - list of extra (Helm) parameters, relevant to the application that needs to be deployed. This would allow for example to set the size of the volumes, enable/disable features, etc... Each parameter would be passed as a separate column in the CSV file, perhaps with some uniform prefix, which is to be defined.

## Notifications

We should also think about the notifications which are sent to the users once the application is deployed. Currently the user is sent an email message containing the name of the application which was deployed, the domain in which it was deployed and an URL to access it (screenshot available below).
![image](https://user-images.githubusercontent.com/4062681/219039382-9c2b971d-c837-4bac-b3eb-1790687f201f.png)

It would be beneficial if this notification message is extended with extra information:
- customizable parameters used during the deployment process (passed via the CSV file or used with default values)
- access information (username, password, IP addresses...)

The idea is to be provide all the necessary information for accessing the application directly in the email.

**Comment by llopat** [comment]
Few questions that come to my mind:
- do we need some application instance name column?
- do we plan to allow for deployment of multiple different applications via a single csv? This would make things more complex, mainly thinking about the custom deployment parameters to be passed in dedicated columns
- do you think that dynamic domain creation for users via the "application deployment csv" brings much value for the user requesting the deployments? Splitting this into two steps (allowing domain creation from a different "csv") might make it a bit less complex

**Comment by korvoj** [comment]
> do we need some application instance name column?

You are right. I was thinking that the user deploying the applications would only need to pass in the ingress hostname as a custom parameter, but I guess it will be good to include both as optional parameters (both the app instance name and the ingress hostname). If the application name is not passed it can be generated automatically (perhaps reuse the existing mechanism for naming deployments - `$app_name-$id`, I leave this up to you but please keep it descriptive and not completely random such as an UUID). I will update my initial comment with this info, as to have everything in one place.

> do we plan to allow for deployment of multiple different applications via a single csv? This would make things more complex, mainly thinking about the custom deployment parameters to be passed in dedicated columns

No, let's limit this feature to bulk deployment of application instances for a single application per CSV file and see if we need something more advanced in the future.

> do you think that dynamic domain creation for users via the "application deployment csv" brings much value for the user requesting the deployments? Splitting this into two steps (allowing domain creation from a different "csv") might make it a bit less complex

If the implementation process would be simpler in case these are separated into two different processes, then by all means to it that way. Perhaps there would be another GitHub issue for describing this "bulk user creation" feature, but for now I would just like to suggest that we think about how we can implement it to fail gracefully in certain edge cases. It would be very nice if the behavior is something like this:
- a list of users is uploaded with a set of required fields (to be determined) including the domain group(s) in which they should join
- the missing domain groups are created (what would be some sane defaults here, e.g. for the initial list of allowed applications?)
- the missing users are created, along with a corresponding domain to which they are assigned as domain managers. Each domain is joined to the respective domain groups as specified in the CSV file.
- in case a user already exists with the same ID (e.g. username), the user creation and domain creation processes are skipped, but we still assign their existing domain to the specified domain groups (UNION with existing domain groups which they are already part of).

**Comment by llopat** [comment]
@korvoj having in mind that we agreed on simplified approach of having instances of only a single application deployed with a single CSV, could we remove the `application name` column from the file and pass it separately? This would slightly simplify the CSV but not sure if it might break the foreseen user process for preparing the CSV and triggering the deployment.

**Comment by korvoj** [comment]
@llopat, yes, I think this is fine. So there would a separate (independent from the CSV) input field in the wizard where the user doing the deployment can choose the desired application?

**Comment by llopat** [comment]
@korvoj yes, I was indeed thinking about such dedicated field or even selector apart from the CSV upload form

**Comment by llopat** [comment]
@korvoj not sure if it was officially agreed already, I suppose we should allow for multiple instances of given application to be deployed in a single domain by adding multiple entries with given domain name in the CSV, correct?

**Comment by korvoj** [comment]
@llopat, from the perspective of the virtual lab use-case, I think this is not required. However, it might be beneficial in some other scenarios, so as to allow as much flexibility as possible, it would be nice if it is implemented that way.

**Comment by llopat** [comment]
Topic to be further explored in the scope of v1.7.0
